src/unitxt/metrics.py

```diff
@@ -400,31 +699,180 @@ def process(self, stream: Stream, stream_name: Optional[str] = None) -> Generato
+                )
+
+        # a list where each element is a group (not an instance), and the dict corresponds to
+        # the aggregate scores of each field within that group
+        return [
+            {
+                "score": {
+                    "instance": {
+                        field_name: group_aggregation_func(scores[field_name])
+                        for field_name in field_names
+                    }
+                }
+            }
+            for scores in group_to_instance_scores.values()
+        ]
+
+    def get_group_aggregated_instance_scores_by_subgroups(
+        self,
+        instances: List[dict],
+        field_names: List[str],
+        group_aggregation_func,
+        subgroup_field: str,
+        baseline_name,
+    ):
+        """Return a list of group aggregation function value for group_mean_subgroup_comparison reduction.
+
+        Args:
+            instances: List of observation instances with instance-level scores (fields) computed.
+            field_names: List of instance score names in each instance to apply the aggregation function.
+            group_aggregation_func: Callable aggregation function with arguments baseline_scores and other_scores.
+            subgroup_field: name of field in instance additional_inputs that contains the subgroup identifier
+            baseline_name: value of subgroup_field that indicates a score belongs in baseline_scores.
+
+        Returns:
+            List of dicts, each corresponding to a group of instances (defined by grouping_field)
+        """
+        from collections import defaultdict
+
+        # two-level defaultdict: first is the grouping, second is the field name
+        # first list is instance scores for baseline group, second is for comparison group
+        group_to_instance_scores = defaultdict(lambda: defaultdict(lambda: [[], []]))
 
         for instance in instances:
-            yield instance
+            additional_inputs = instance["additional_inputs"]
+            for cc in [self.grouping_field, subgroup_field]:
+                if cc not in additional_inputs:
+                    raise ValueError(
+                        f"Missing '{cc}' from instance {instance}. "
+                        f"This field is required for group based metric computation."
+                    )
+            group_key = additional_inputs[self.grouping_field]
+            # indicator if is in the baseline group or not
+            is_baseline = str(additional_inputs[subgroup_field]) == baseline_name
+            # convert True (baseline) to 0, and False (others) to 1, store in respective groups
+            idx = int(not is_baseline)
+
+            for field_name in field_names:
+                group_to_instance_scores[group_key][field_name][idx].append(
+                    instance["score"]["instance"][field_name]
+                )
+
+        # now for each group, take the aggregation function, comparing others to baseline
+        return [
```

Comment:
Here, i things it would be best to just return a list of group scores (where each list item is a dict with mapping from  `field_name` to the score value).
I realize that would mean making other changes since the current code expects the score maps to be within `instance["score"]["instance"]`.. but since we are now dealing with groups that convention becomes very confusing..

Reply:
I think what I did is okay because it allows the code to be used for both reductions.  Plus the scores are only used internally and not returned in the instance list.